Better unchecked warnings.

paulp opened the pull request.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/583/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1290/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/583/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1290/

paulp commented:
PLS REBUILD ALL

paulp commented:
Review by @adriaanm since he seems to be all back and stuff.  Tell me I missed some massively easier way.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/585/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1292/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/585/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1292/

retronym commented on src/compiler/scala/tools/nsc/typechecker/Checkable.scala:
```diff
@@ -0,0 +1,247 @@
+      """.stripMargin.trim
+
+      val msg = List(facts, outcomes) mkString ("\n", "\n", "\n")
+      log(msg)
```
Create the message inside the lazy arg to `log`.

paulp replied:
Ugh, oops. I was trying to get this in before I had to bail yesterday.

gkossakowski commented:
Just to set right expectations on @adriaanm's availability: he is back but rather busy with non-work stuff so I wouldn't put too much pressure on him.

Other than that it looks like a fairly big patch. I hope it can get lgtm by Friday.

retronym commented:
To make this a bit easier to review, could please you split the commit into two pieces: one that extracts the `Checkable` trait, and a second that implements the behavioural changes?

paulp commented:
Okay, forget @adriaanm, I'll attempt to meet @retronym's review demands.

jsuereth commented:
BTW - We had a discussion at the Scala meeting on releasing.  I think this is best suited to target 2.10.1 (which should be out in 4 months + N weeks).   We'll lock the date for 2.10.1 down soon.

Also, we haven't firmed up where 2.10.1 dev will happen yet.   I'll get back to you on whether we spin up a temporary branch for the next month, or just put things in master.

paulp commented:
You should turn off -unchecked by default if this isn't going in.

paulp committed:
Move checkCheckable out of Infer.

Preparing for separate file with checkability logic.

retronym commented:
I would like to see this one given a chance. If it issues a spurious warning that needs be refined in RC 2 or 3, what's the harm? It won't block people using the RC for other means.

paulp committed:
Added exists3 to fast collections.

paulp commented:
I have made the patch much stronger.

paulp commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/601/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1308/

paulp commented:
files/run/t6344.scala is going to fail because that checkfile update wasn't in.  If that's it, let's pretend it passed.

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1308/

paulp commented:
Indeed:

```
build.xml:2649: java.lang.RuntimeException: Test suite finished with 1 case failing:
test/files/run/t6344.scala [FAILED]
```

paulp committed:
Moved isVariableName to StdNames where it belong.s

paulp committed:
Pass a more precise type to checkCheckable when it's available.

Without it, one cannot assess the checkability of any
aspect of the pattern for which static verifiability
depends on knowledge of the scrutinee.

paulp committed:
Improvements to unchecked warnings.

Closes SI-6275, SI-5762.

The comment says is better than I can.

/** On pattern matcher checkability:
 *
 *  Consider a pattern match of this form: (x: X) match { case _: P => }
 *
 *  There are four possibilities to consider:
 *     [P1] X will always conform to P
 *     [P2] x will never conform to P
 *     [P3] X <: P if some runtime test is true
 *     [P4] X cannot be checked against P
 *
 *  The first two cases correspond to those when there is enough static
 *  information to say X <: P or that !(X <: P) for all X and P.
 *  The fourth case includes unknown abstract types or structural
 *  refinements appearing within a pattern.
 *
 *  The third case is the interesting one.  We designate another type, XR,
 *  which is essentially the intersection of X and |P|, where |P| is
 *  the erasure of P.  If XR <: P, then no warning is emitted.
 *
 *  Examples of how this info is put to use:
 *  sealed trait A[T] ; class B[T] extends A[T]
 *    def f(x: B[Int]) = x match { case _: A[Int] if true => }
 *    def g(x: A[Int]) = x match { case _: B[Int] => }
 *
 *  `f` requires no warning because X=B[Int], P=A[Int], and B[Int] <:< A[Int].
 *  `g` requires no warning because X=A[Int], P=B[Int], XR=B[Int], and B[Int] <:< B[Int].
 *      XR=B[Int] because a value of type A[Int] which is tested to be a B can
 *      only be a B[Int], due to the definition of B (B[T] extends A[T].)
 *
 *  This is something like asSeenFrom, only rather than asking what a type looks
 *  like from the point of view of one of its base classes, we ask what it looks
 *  like from the point of view of one of its subclasses.
 */

paulp committed:
Additional new tests for unchecked warnings.

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/601/

paulp commented:
Reopening a new one.
